cli/inspector.rs

```diff
@@ -0,0 +1,549 @@
+/// inspector channel on the isolate thread to the websocket that is owned by
+/// the inspector server.
+type SessionToFrontendTx = mpsc::UnboundedSender<ws::Message>;
+/// Owned by the inspector server. Messages arriving on this channel, coming
+/// from the inspector session on the isolate thread are forwarded over the
+/// websocket to the devtools frontend.
+type SessionToFrontendRx = mpsc::UnboundedReceiver<ws::Message>;
+
+/// Stored in a UUID hashmap, used by WS server. Clonable.
+#[derive(Clone)]
+struct InspectorInfo {
+  uuid: Uuid,
+  frontend_to_inspector_tx: FrontendToInspectorTx,
+  inspector_handle: DenoInspectorHandle,
+}
+
+/// Owned by GlobalState.
+pub struct InspectorServer {
+  address: SocketAddrV4,
+  thread_handle: Option<std::thread::JoinHandle<()>>,
+  server_msg_tx: Option<ServerMsgTx>,
+}
+
+impl InspectorServer {
+  pub fn new(host: &str, brk: bool) -> Self {
+    if brk {
+      todo!("--inspect-brk not yet supported");
+    }
+    let address = host.parse::<SocketAddrV4>().unwrap();
+    let (server_msg_tx, server_msg_rx) = mpsc::unbounded_channel::<ServerMsg>();
+    let thread_handle = std::thread::spawn(move || {
+      crate::tokio_util::run_basic(server(address, server_msg_rx));
+    });
+    Self {
+      address,
+      thread_handle: Some(thread_handle),
+      server_msg_tx: Some(server_msg_tx),
+    }
+  }
+
+  /// Each worker/isolate to be debugged should call this exactly one.
+  /// Called from worker's thread
+  pub fn add_inspector(
+    &self,
+    isolate: &mut deno_core::Isolate,
+  ) -> Box<DenoInspector> {
+    let deno_core::Isolate {
+      v8_isolate,
+      global_context,
+      ..
+    } = isolate;
+    let v8_isolate = v8_isolate.as_mut().unwrap();
+
+    let mut hs = v8::HandleScope::new(v8_isolate);
+    let scope = hs.enter();
+    let context = global_context.get(scope).unwrap();
+
+    let server_msg_tx = self.server_msg_tx.as_ref().unwrap().clone();
+    let address = self.address;
+    let (frontend_to_inspector_tx, frontend_to_inspector_rx) =
+      mpsc::unbounded_channel::<FrontendToInspectorMsg>();
+    let uuid = Uuid::new_v4();
+
+    let inspector = crate::inspector::DenoInspector::new(
+      scope,
+      context,
+      frontend_to_inspector_rx,
+    );
+
+    eprintln!(
+      "Debugger listening on {}",
+      websocket_debugger_url(address, &uuid)
+    );
+
+    server_msg_tx
+      .send(ServerMsg::AddInspector(InspectorInfo {
+        uuid,
+        frontend_to_inspector_tx,
+        inspector_handle: DenoInspectorHandle::new(
+          &inspector,
+          v8_isolate.thread_safe_handle(),
+        ),
+      }))
+      .unwrap_or_else(|_| {
+        panic!("sending message to inspector server thread failed");
+      });
+
+    inspector
+  }
+}
+
+impl Drop for InspectorServer {
+  fn drop(&mut self) {
+    self.server_msg_tx.take();
+    self.thread_handle.take().unwrap().join().unwrap();
+    panic!("TODO: this drop is never called");
```

Comment:
Should this be fixed before landing?

Reply:
This is a rusty_v8 bug. It’s a little strange because a panic would indicate proper operation. I think we should keep it until this is fixed.